core/trino-main/src/main/java/io/trino/sql/planner/CachingSubExpressionExtractor.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import io.trino.sql.tree.Expression;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Extracts, caches and returns the stream of all expression subtrees within an Expression, including Expression itself
+ */
+public final class CachingSubExpressionExtractor
+{
+    private final Map<Expression, List<Expression>> subExpressionsCache = new HashMap<>();
```

Comment:
This could be called just `cache` for conciseness without loss of clarity.

Reply:
done